### PR TITLE
Scripts for automated dbSNP schema generation

### PR DESCRIPTION
**Pull request by spiekos**
The scripts used to generate the schema for the enumerations for data sources of allele frequencies and the properties of alternative IDs for genetic variants.

### CODE REVIEW

File: scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+'''
+Author:  Samantha Piekos
+Date:  09/30/20
+Name:  format_dbSNP_GevVarSource_enum_schema.py
+Description: Write all unique population sources recording genetic variant frequencies for a .vcf
```
Comment (pradh) [inline scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py:5]: nit: It's generally good practice to keep the char width to 80 chars.

File: scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import sys
+
+def collect_freq_pop(freq_pops, set_freq_pop):
+	'''
+	For each genetic variant instance add all populations for its frequencies to a set.
+	@freq_pops		string of all populations recording frequencies for a genetic variant instance
+	@set_freq_pop	set of all populations recording genetic variant frequencies
+	@return			set of all populations recording genetic variant frequencies
+	'''
+	pops = freq_pops.split('|')
```
Comment (pradh) [inline scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py:21]: What is the indent spacing in the script?  I think this repo uses 4 spaces as one indent.

File: scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	for item in pops:
+		if item == 'stonian':
+			item = 'E' + item
+		if item != '.':
+			item = item.capitalize()  # make sure first character of the name is uppercase
+			set_freq_pop.add(item.split(':')[0])
+	return(set_freq_pop)
```
Comment (pradh) [inline scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py:28]: nit: return set_freq_pop

File: scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	'''
+	Compile list of all populations for genetic variant frequencies.
+	@file_input	input .vcf or .txt file of genetic variants
+	@return		unique list of all populations recording genetic variant frequencies
+	'''
+	set_freq_pop = set()
+	f = open(file_input, mode='r')
+	for line in f:
+		line = line.strip('\r\n').split('\t')
+		values = line[7].split(';')
```
Comment (pradh) [inline scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py:40]: comment the significance of 7?

File: scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		for item in values:
+			if item.startswith("FREQ="):
+				set_freq_pop = collect_freq_pop(item.strip('FREQ='), set_freq_pop)
+	return(list(set_freq_pop))
```
Comment (pradh) [inline scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py:44]: nit: return list(set_freq_pop)

File: scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	'''
+	For each genetic variant instance add all databases with an alternative ID for it to a set.
+	@db_IDs			string of all databases with alternative IDs for a given genetic variant instance
+	@set_alt_db_IDs	set of all databases with alternative IDs for the genetic variant
+	@return			set of all databases with alternative IDs for genetic variants
+	'''
+	list_existing_ID_prop = ['dbvar', 'clinvar', 'pharmgkb']
+	pops = db_IDs.split(',')
+	for item in pops:
+		if item.lower() in list_existing_ID_prop or len(item) == 0:  # remove those with existing schema
```
Comment (pradh) [inline scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py:43]: is item guaranteed to not have leading/trailing spaces? if not, its good practice to do a item.strip()

File: scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	Write all populations recording genetic variant frequencies as enums of class
+	GenVarSourceEnum to an output mcf file.
+	@file_output	mcf output file recording all databases with alternative IDs for genetic variants as properties of GeneticVariant
+	@return			unique list of all populations recording genetic variant frequencies
+	'''
+	w = open(file_output, mode='w')
+	w.write('# Generated by script format_dbSNP_alt_ID_database_property_schema.py\n\n')
+	for item in list_alt_db_IDs:
+		w.write('Node: dcid:'+ item + '_ID\n')
+		w.write('name: "' + item + '_ID"\n')
```
Comment (pradh) [inline scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py:94]: Its good to not have '_' in the property name.

**Review by pradh — COMMENTED** [review]
Please consider renaming the PR title to something more meaningful.

To confirm, is this the code to generate https://github.com/datacommonsorg/data/pull/324 ?  If so, the main comment is to remove extraneous non-schema-friendly characters like '_', ')', '(' from property names.  Otherwise, just some minor code readability/convention comments.

**Review by spiekos — COMMENTED** [review]
Addressed all of your comments.

File: scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py (hunk 6 above)
Comment (spiekos) [reply scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py]: Strip was done on line 46 of the script

File: scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py [hunk 8]
```diff
@@ -2,19 +2,36 @@
...
+	'''
+	Convert string to camel case.
+
+	@l		list of components originally seperated by '_' to be reconstructed
+			in camel case
+	@return	camel case string
+	'''
+	camel_item = ""
+	for i in l:
+		camel_item = camel_item + i[0].upper() + i[1:]
```
Comment (pradh) [inline scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py:26]: nit: can use i.capitalize() ?

File: scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py [hunk 9]
```diff
@@ -23,32 +40,39 @@ def collect_freq_pop(freq_pops, set_freq_pop):
...
 			item = 'E' + item
 		if item != '.':
-			item = item.capitalize()  # make sure first character of the name is uppercase
+			# make sure first character of the name is uppercase
+			item = item.capitalize()
+			if '_' in item:
+				item = format_name(item.split('_'))
 			set_freq_pop.add(item.split(':')[0])
-	return(set_freq_pop)
+	return set_freq_pop
```
Comment (pradh) [inline scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py:48]: nit: is there a trailing space?

File: scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py [hunk 10]
```diff
@@ -2,56 +2,94 @@
...
+	Converts string with words seperated by '_' to camel case.
+
+	@item	string whose words are separated by '_'
+	@return	string that is in camel case
+	"""
+	camel_item = ""
+	l = item.split('_')
+	if len(l) > 2:
+		for i in l:
+			if len(i) > 2:
```
Comment (pradh) [inline scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py:27]: should it be > 1? Or do you specifically only want to capitalize for 3 characters or more?

File: scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py [hunk 11]
```diff
@@ -2,56 +2,94 @@
...
+				i = i.lower()
+				camel_item = camel_item + i[0].upper() + i[1:]
```
Comment (pradh) [inline scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py:29]: nit: i.capitalize()?

**Review by pradh — APPROVED** [review]
Thanks for the changes. Some minor comments. Please feel free to merge after.

File: scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py (hunk 8 above)
Comment (spiekos) [reply scripts/dbSNP/format_dbSNP_GenVarSource_enum_schema.py]: changed to i.capitalize()

File: scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py (hunk 10 above)
Comment (spiekos) [reply scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py:41]: There is weird formatting here in which there's _ used when it shouldn't be resulting in an error thrown. if it's greater than 2 this issue is avoided.

File: scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py (hunk 11 above)
Comment (spiekos) [reply scripts/dbSNP/format_dbSNP_alt_ID_database_property_schema.py]: done

**Comment by spiekos** [comment]
\gcbrun

**Comment by spiekos** [comment]
\gcbrun

**Pull request merged**